Crash on editing Markdown file

### Summary

It crashes.

### Reproduction Steps

I tried this:
1. `RUST_BACKTRACE=1 hx -vv bug.md`
2. Enter the following text:
````
```bash
#!/usr/bin/env bash
```
````
3. Copy "shebang" before the first line: `ggjxyggP`.

I expected to get this text:

````
#!/usr/bin/env bash
```bash
#!/usr/bin/env bash
```
````

Instead, this happened:
```
thread 'main' panicked at 'assertion failed: byte_idx <= slice.len_bytes()', helix-core/src/graphemes.rs:330:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:48:5
   3: helix_core::graphemes::is_grapheme_boundary_byte
             at ./helix-core/src/graphemes.rs:330:5
   4: helix_core::graphemes::ensure_grapheme_boundary_next_byte
             at ./helix-core/src/graphemes.rs:290:12
   5: helix_term::ui::editor::EditorView::doc_syntax_highlights::{{closure}}
             at ./helix-term/src/ui/editor.rs:313:51
   6: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:306:13
   7: core::option::Option<T>::map
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:929:29
   8: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/iter/adapters/map.rs:103:9
   9: <alloc::boxed::Box<I,A> as core::iter::traits::iterator::Iterator>::next
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1874:9
  10: <helix_core::syntax::Merge<I> as core::iter::traits::iterator::Iterator>::next
             at ./helix-core/src/syntax.rs:2373:35
  11: <alloc::boxed::Box<I,A> as core::iter::traits::iterator::Iterator>::next
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1874:9
  12: <helix_term::ui::document::StyleIter<H> as core::iter::traits::iterator::Iterator>::next
             at ./helix-term/src/ui/document.rs:49:33
  13: helix_term::ui::document::render_text
             at ./helix-term/src/ui/document.rs:264:26
  14: helix_term::ui::document::render_document
             at ./helix-term/src/ui/document.rs:106:5
  15: helix_term::ui::editor::EditorView::render_view
             at ./helix-term/src/ui/editor.rs:188:9
  16: <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::render
             at ./helix-term/src/ui/editor.rs:1412:13
  17: helix_term::compositor::Compositor::render
             at ./helix-term/src/compositor.rs:170:13
  18: helix_term::application::Application::render::{{closure}}
             at ./helix-term/src/application.rs:289:9
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  20: helix_term::application::Application::handle_terminal_events::{{closure}}
             at ./helix-term/src/application.rs:671:26
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  22: helix_term::application::Application::event_loop_until_idle::{{closure}}
             at ./helix-term/src/application.rs:332:55
  23: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  24: helix_term::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:306:57
  25: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  26: helix_term::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:1133:38
  27: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  28: hx::main_impl::{{closure}}
             at ./helix-term/src/main.rs:157:53
  29: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  30: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/park.rs:283:63
  31: tokio::runtime::coop::with_budget
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/coop.rs:107:5
  32: tokio::runtime::coop::budget
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/coop.rs:73:5
  33: tokio::runtime::park::CachedParkThread::block_on
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/park.rs:283:31
  34: tokio::runtime::context::BlockingRegionGuard::block_on
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/context.rs:315:13
  35: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/scheduler/multi_thread/mod.rs:66:9
  36: tokio::runtime::runtime::Runtime::block_on
             at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.28.2/src/runtime/runtime.rs:304:45
  37: hx::main_impl
             at ./helix-term/src/main.rs:159:5
  38: hx::main
             at ./helix-term/src/main.rs:38:21
  39: core::ops::function::FnOnce::call_once
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-06-23T17:42:32.667 helix_loader [DEBUG] runtime dir: /home/user/t/helix/runtime
2023-06-23T17:42:32.686 helix_view::clipboard [DEBUG] Using wl-copy+wl-paste to interact with the system and selection (primary) clipboard
2023-06-23T17:42:32.728 helix_vcs [INFO] file is untracked
2023-06-23T17:42:32.728 helix_vcs [INFO] failed to open diff base for /home/user/t/helix/bug.md
2023-06-23T17:42:32.731 helix_view::editor [ERROR] Failed to initialize the language servers for `source.md` { cannot find binary path }
2023-06-23T17:42:32.731 helix_view::editor [DEBUG] editor status: Loaded 1 file.
2023-06-23T17:42:32.733 helix_tui::backend::crossterm [DEBUG] The keyboard enhancement protocol is not supported in this terminal (checked in 383.016µs)
2023-06-23T17:42:32.735 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:33.088 helix_term::application [DEBUG] received editor event: IdleTimer
2023-06-23T17:42:34.445 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:34.842 helix_term::application [DEBUG] received editor event: IdleTimer
2023-06-23T17:42:34.847 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:35.307 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:35.489 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:35.813 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:36.211 helix_term::application [DEBUG] received editor event: IdleTimer
2023-06-23T17:42:36.216 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-06-23T17:42:36.533 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
...
```

</details>


### Platform

Linux

### Terminal Emulator

alacritty 0.12.1 (5fdfd47f)

### Helix Version

23.05-168-g2c5288da

--- a/helix-core/src/syntax.rs
+++ b/helix-core/src/syntax.rs
@@ -1136,6 +1136,7 @@ impl Syntax {
                     RopeProvider(source_slice),
                 );
                 let mut injections = Vec::new();
+                let mut last_injection_end = 0;
                 for mat in matches {
                     let (injection_capture, content_node, included_children) = layer
                         .config
@@ -1155,6 +1156,10 @@ impl Syntax {
                                 intersect_ranges(&layer.ranges, &[content_node], included_children);
 
                             if !ranges.is_empty() {
+                                if content_node.start_byte() < last_injection_end {
+                                    continue;
+                                }
+                                last_injection_end = content_node.end_byte();
                                 injections.push((config, ranges));
                             }
                         }

--- a/helix-core/src/syntax.rs
+++ b/helix-core/src/syntax.rs
@@ -1135,12 +1135,37 @@ impl Syntax {
                     layer.tree().root_node(),
                     RopeProvider(source_slice),
                 );
+                let mut combined_injections = vec![
+                    (None, Vec::new(), IncludedChildren::default());
+                    layer.config.combined_injections_patterns.len()
+                ];
                 let mut injections = Vec::new();
                 let mut last_injection_end = 0;
                 for mat in matches {
                     let (injection_capture, content_node, included_children) = layer
                         .config
                         .injection_for_match(&layer.config.injections_query, &mat, source_slice);
+
+                    // in case this is a combined injection save it for more processing later
+                    if let Some(combined_injection_idx) = layer
+                        .config
+                        .combined_injections_patterns
+                        .iter()
+                        .position(|&pattern| pattern == mat.pattern_index)
+                    {
+                        let entry = &mut combined_injections[combined_injection_idx];
+                        if injection_capture.is_some() {
+                            entry.0 = injection_capture;
+                        }
+                        if let Some(content_node) = content_node {
+                            if content_node.start_byte() >= last_injection_end {
+                                entry.1.push(content_node);
+                                last_injection_end = content_node.end_byte();
+                            }
+                        }
+                        entry.2 = included_children;
+                        continue;
+                    }
 
                     // Explicitly remove this match so that none of its other captures will remain
                     // in the stream of captures.
@@ -1166,43 +1191,13 @@ impl Syntax {
                     }
                 }
 
-                // Process combined injections.
-                if let Some(combined_injections_query) = &layer.config.combined_injections_query {
-                    let mut injections_by_pattern_index =
-                        vec![
-                            (None, Vec::new(), IncludedChildren::default());
-                            combined_injections_query.pattern_count()
-                        ];
-                    let matches = cursor.matches(
-                        combined_injections_query,
-                        layer.tree().root_node(),
-                        RopeProvider(source_slice),
-                    );
-                    for mat in matches {
-                        let entry = &mut injections_by_pattern_index[mat.pattern_index];
-                        let (injection_capture, content_node, included_children) = layer
-                            .config
-                            .injection_for_match(combined_injections_query, &mat, source_slice);
-                        if injection_capture.is_some() {
-                            entry.0 = injection_capture;
-                        }
-                        if let Some(content_node) = content_node {
-                            entry.1.push(content_node);
-                        }
-                        entry.2 = included_children;
-                    }
-                    for (lang_name, content_nodes, included_children) in injections_by_pattern_index
-                    {
-                        if let (Some(lang_name), false) = (lang_name, content_nodes.is_empty()) {
-                            if let Some(config) = (injection_callback)(&lang_name) {
-                                let ranges = intersect_ranges(
-                                    &layer.ranges,
-                                    &content_nodes,
-                                    included_children,
-                                );
-                                if !ranges.is_empty() {
-                                    injections.push((config, ranges));
-                                }
+                for (lang_name, content_nodes, included_children) in combined_injections {
+                    if let (Some(lang_name), false) = (lang_name, content_nodes.is_empty()) {
+                        if let Some(config) = (injection_callback)(&lang_name) {
+                            let ranges =
+                                intersect_ranges(&layer.ranges, &content_nodes, included_children);
+                            if !ranges.is_empty() {
+                                injections.push((config, ranges));
                             }
                         }
                     }
@@ -1565,7 +1560,7 @@ pub struct HighlightConfiguration {
     pub language: Grammar,
     pub query: Query,
     injections_query: Query,
-    combined_injections_query: Option<Query>,
+    combined_injections_patterns: Vec<usize>,
     highlights_pattern_index: usize,
     highlight_indices: ArcSwap<Vec<Option<Highlight>>>,
     non_local_variable_patterns: Vec<bool>,
@@ -1681,26 +1676,15 @@ impl HighlightConfiguration {
             }
         }
 
-        let mut injections_query = Query::new(language, injection_query)?;
-
-        // Construct a separate query just for dealing with the 'combined injections'.
-        // Disable the combined injection patterns in the main query.
-        let mut combined_injections_query = Query::new(language, injection_query)?;
-        let mut has_combined_queries = false;
-        for pattern_index in 0..injections_query.pattern_count() {
-            let settings = injections_query.property_settings(pattern_index);
-            if settings.iter().any(|s| &*s.key == "injection.combined") {
-                has_combined_queries = true;
-                injections_query.disable_pattern(pattern_index);
-            } else {
-                combined_injections_query.disable_pattern(pattern_index);
-            }
-        }
-        let combined_injections_query = if has_combined_queries {
-            Some(combined_injections_query)
-        } else {
-            None
-        };
+        let injections_query = Query::new(language, injection_query)?;
+        let combined_injections_patterns = (0..injections_query.pattern_count())
+            .filter(|&i| {
+                injections_query
+                    .property_settings(i)
+                    .iter()
+                    .any(|s| &*s.key == "injection.combined")
+            })
+            .collect();
 
         // Find all of the highlighting patterns that are disabled for nodes that
         // have been identified as local variables.
@@ -1749,7 +1733,7 @@ impl HighlightConfiguration {
             language,
             query,
             injections_query,
-            combined_injections_query,
+            combined_injections_patterns,
             highlights_pattern_index,
             highlight_indices,
             non_local_variable_patterns,

